### PR TITLE
azurerm_network_interface & azurerm_arm_loadbalancer: fixing d.Set() set errors and some additional validation

### PR DESCRIPTION
**Pull request by katbyte**
a result of investigating #1318

### CODE REVIEW

File: azurerm/helpers/validate/azure.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	v, ok := i.(string)
+	if !ok {
+		errors = append(errors, fmt.Errorf("expected type of %q to be string", k))
+		return
+	}
+
+	//todo, match subscription/{guid} and provider/name.name perhaps?
+	if matched := regexp.MustCompile(`(/subscriptions/)|(/providers/)`).Match([]byte(v)); !matched {
+		errors = append(errors, fmt.Errorf("azure resource ID %q should start with with '/subscriptions/{subscriptionId}' or '/providers/{resourceProviderNamespace}/'", k))
+	}
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/helpers/validate/azure.go]: `tenants` are also a possible prefix too e.g. `/tenants/000/subscriptions/000/...`

File: azurerm/helpers/validate/network_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			Ip:     "ab:cd:ef:AB:CD:EF",
+			Errors: 0,
+		},
+	}
+
+	for _, tc := range cases {
+		_, errors := Ip4Address(tc.Ip, "test")
+
+		if len(errors) < tc.Errors {
+			t.Fatalf("Expected AzureResourceId to have an error for '%q'", tc.Ip)
```
Comment (tombuildsstuff) [inline azurerm/helpers/validate/network_test.go]: _minor_ technically `'%q'` can be just `%q`

File: azurerm/resource_arm_loadbalancer.go [hunk 3]
```diff
@@ -34,56 +36,81 @@ func resourceArmLoadBalancer() *schema.Resource {
...
 				Optional: true,
 				MinItems: 1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"name": {
-							Type:     schema.TypeString,
-							Required: true,
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.StringNotEmpty,
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_loadbalancer.go]: there's a generic validation function in Core for "no default value" which may be better to use?

Comment (katbyte) [reply azurerm/resource_arm_loadbalancer.go]: I didn't know about that one! `validation.NoZeroValues` is perfect.

File: azurerm/helpers/azure/resourceid.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	// It is OK not to have a provider in the case of a resource group
+	if provider, ok := componentMap["providers"]; ok {
+		idObj.Provider = provider
+		delete(componentMap, "providers")
+	}
+
+	return idObj, nil
+}
+
+func composeAzureResourceID(idObj *ResourceID) (id string, err error) {
```
Comment (tombuildsstuff) [inline azurerm/helpers/azure/resourceid.go]: I don't see anything using this apart from the Tests, in which case can we remove this?

File: azurerm/resource_arm_loadbalancer.go [hunk 5]
```diff
@@ -34,56 +37,81 @@ func resourceArmLoadBalancer() *schema.Resource {
...
 						},
 
 						"private_ip_address": {
-							Type:     schema.TypeString,
-							Optional: true,
-							Computed: true,
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: azure.ValidateResourceId,
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_loadbalancer.go]: this is an IP Address, not a Resource ID?

File: azurerm/helpers/azure/resourceid.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	ResourceGroup  string
+	Provider       string
+	Path           map[string]string
+}
+
+// parseAzureResourceID converts a long-form Azure Resource Manager ID
+// into a ResourceID. We make assumptions about the structure of URLs,
+// which is obviously not good, but the best thing available given the
+// SDK.
+func ParseAzureResourceID(id string) (*ResourceID, error) {
```
Comment (tombuildsstuff) [inline azurerm/helpers/azure/resourceid.go]: are we also going to call this method from the existing method to give us a migration path whilst refactoring?

File: azurerm/resource_arm_loadbalancer.go (hunk 5 above)
Comment (katbyte) [reply azurerm/resource_arm_loadbalancer.go]: Whoops, got copy paste happy it seems

File: azurerm/helpers/azure/resourceid.go (hunk 6 above)
Comment (katbyte) [reply azurerm/helpers/azure/resourceid.go]: Thats a good idea, now threading down to new method

File: azurerm/helpers/azure/resourceid.go (hunk 4 above)
Comment (katbyte) [reply azurerm/helpers/azure/resourceid.go]: I removed the top level one, and kept this and the tests here on the new function

Comment (tombuildsstuff) [reply azurerm/helpers/azure/resourceid.go]: sorry, I meant since this is Private and it's not being used within this Package, I think this is dead code?

**Pull request merged**